Show current wireless LAN status in N900 status menu
by checking if the wl12xx module is loaded.

Also rename button from "Wifi switcher" to "Wireless LAN"
and include the state text in the sub label to match the
look of the built-in bluetooth switcher.

Signed-off-by: Thomas Jarosch <tomj@simonv.com>
diff -u -r -p wifi-switcher-0.0.9.orig/src/wifi-switcher-statusbar.c wifi-switcher-0.0.9/src/wifi-switcher-statusbar.c
--- wifi-switcher-0.0.9.orig/src/wifi-switcher-statusbar.c	2010-02-03 07:07:52.000000000 +0000
+++ wifi-switcher-0.0.9/src/wifi-switcher-statusbar.c	2010-07-16 21:07:45.981322870 +0100
@@ -54,6 +54,40 @@ static void wifi_switcher_class_init (Wi
 	g_type_class_add_private (klass, sizeof (WifiSwitcherPrivate));
 }
 
+/** Check if wifi module is loaded
+    @return One if wifi module is loaded, zero otherwise
+*/
+static int check_wifi_module_loaded()
+{
+	char line[128];
+	FILE *fp = fopen("/proc/modules", "r");
+	if (!fp)
+		return 0;
+
+	while (fgets(line, sizeof(line), fp) != NULL)
+	{
+		if (strstr(line, "wl12xx") == line)
+		{
+			fclose(fp);
+			return 1;
+		}
+	}
+
+	fclose (fp);
+	return 0;
+}
+
+/** Set Wifi button text according to wifi module state
+	@param priv Pointer to WifiSwitcherPrivate structure
+*/
+static void set_wifi_button_text(WifiSwitcherPrivate *priv)
+{
+	const char *subtext = check_wifi_module_loaded() ? "On" : "Off";
+	hildon_button_set_text (HILDON_BUTTON (priv->button),
+				"Wireless LAN",
+				subtext);
+}
+
 static void wifi_switcher_init (WifiSwitcher *plugin)
 {
 	WifiSwitcherPrivate *priv = WIFI_SWITCHER_GET_PRIVATE (plugin);
@@ -64,7 +98,7 @@ static void wifi_switcher_init (WifiSwit
 
 	hildon_button_set_style (HILDON_BUTTON(priv->button), HILDON_BUTTON_STYLE_PICKER);
 	hildon_button_set_image (HILDON_BUTTON(priv->button), image);
-	hildon_button_set_text (HILDON_BUTTON (priv->button), "Wifi Switcher", NULL);
+	set_wifi_button_text(priv);
 	hildon_button_set_alignment (HILDON_BUTTON(priv->button), 0.0, 0.0, 1.0, 1.0);
 
 	g_signal_connect_after (G_OBJECT (priv->button), "clicked", G_CALLBACK (wifi_switcher_cb), plugin);
@@ -74,9 +108,13 @@ static void wifi_switcher_init (WifiSwit
 
 static gboolean run(gpointer plugin)
 {
+	WifiSwitcherPrivate *priv = WIFI_SWITCHER_GET_PRIVATE (plugin);
 	int status =  WEXITSTATUS(system("sudo /usr/sbin/wifi.sh"));
 	GdkPixbuf *pixbuf = NULL;
 
+	/* Update wifi button text */
+	set_wifi_button_text(priv);
+
 	if (status == 2) {
 		/* Put the icon at the status bar*/	
 		pixbuf = gdk_pixbuf_new_from_file_at_size ("/usr/share/icons/hicolor/scalable/hildon/wlan_off.png",
